src/tokens/wrappers/clawback/Clawback.sol

```diff
@@ -0,0 +1,298 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import {IClawback, IClawbackFunctions} from "./IClawback.sol";
+import {IERC721Transfer} from "../../common/IERC721Transfer.sol";
+
+import {IERC1155} from "@0xsequence/erc-1155/contracts/interfaces/IERC1155.sol";
+import {IERC165} from "@0xsequence/erc-1155/contracts/interfaces/IERC165.sol";
+import {ERC1155Metadata} from "@0xsequence/erc-1155/contracts/tokens/ERC1155/ERC1155Metadata.sol";
+import {ERC1155, ERC1155MintBurn} from "@0xsequence/erc-1155/contracts/tokens/ERC1155/ERC1155MintBurn.sol";
+
+import {SafeTransferLib} from "solady/utils/SafeTransferLib.sol";
+
+contract Clawback is ERC1155MintBurn, ERC1155Metadata, IClawback {
```

Comment:
Few notes ; 

1. Looks like this is one wrapper contract for all token contracts. This is going to make it hard for indexing. I think 1 clawback wrapper per contract is cleaner, with one set of operator and such. 

2. Ideally most games launching on sequence will have their tokens natively support clawback system. meaning they don't need to wrap in a separate contract. Having one erc-20 contract vs 2 is nice because they don't need to change any of the contracts. It would work on Opensea, Sequence marketplace, etc. without any changes. It's just that the transfer would revert if opensea wasn't whitelisted for this user. The current implementation is not backward compatible with any existing contract.  I understand for existing collections and contracts wrapping is required, but we still need a native solution as outlined [here](https://github.com/0xsequence/internal-docs/blob/master/products/payments/clawback_token.md#--%EF%B8%8F-overview)

3. How does the `templateTransferers` work? Say you wanted to let people use ERC-20 for our Sequence Marketplace. How would this work? They would put the sequence orderbook contract as a transferer? How would our marketplace pull the funds and unwrap them to complete the order? I understand this would be easier with native clawback support, but not sure I understand the flow for non-native ones.

4. How does one force a token to be unwrappable manually? Looks like there's not way to speed up an unvesting as per outlined in the [set_maturity function section](https://github.com/0xsequence/internal-docs/blob/master/products/payments/clawback_token.md#---feature)

 5. Why only allow the owner to unwrap? Being in a "clawback state" is not a good thing and there could be third parties that will help with unwrapping on behalf of users. Native implementations shouldn't need this as the un-wrapping should be "on-demand", but for the non-native implementations we should let third parties unwrap on behalf of users if maturity has passed

Reply:
For reference, the implementation is based of this technical spec: https://hackmd.io/Ucg-sUV_THOWwG3kscpO9g
I wasn't across the spec you've linked and there are some differences between the two. 

1. 
To support indexing I've added `tokenAddr` as a parameter in wrap, unwrap and clawback events. (Just added `indexed` flag as well). 
With 1 wrapper per contract, we still have the problem of multiple parties wanting to create their own vesting rules (e.g. Project team, Fiat onramps, Players in game). For fiat onramp providers in particular, having a single contract with a single template that is used for all tokens would make things easier as they can generalise the clawback solution and management, which would be consistent across collections. 
One thing to consider with the current approach is that if the clawback contract is shown in a whitelabelled marketplace the tokens from other projects would be shown. This can be avoided with some filtering on our end. I think custom handling of the clawback contract in marketplace would be useful anyway. Buying unvested tokens is a big risk and we should label those tokens accordingly. 

2. 
Wrapped instances need to be non-fungible so that each wrapped instance can be tracked. If we wrapped within an ERC-20, the fungible nature will prevent the ability to track vesting status. It's important to be able to differentiate which tokens are under vesting rules as that will impact the value of those tokens (as per the marketplace example). ERC-20 doesn't support multiple tokens on a single contract. Mixing ERC-20/721/1155 standards on a single contract will cause integration issues. 
We could make a native implementations for 721 and 1155. However ERC-1155 has similar problems to ERC-20 within a tokenId and there are tokenId collision risks. 
The current implementation will accept any token and is backwards and forwards compatible. 

3. 
See the `Custom` under [Transferability in the tech spec](https://hackmd.io/Ucg-sUV_THOWwG3kscpO9g#Transferability). 
For Sequence marketplace the flow would be:
1. Vesting template is created. 
2. Marketplace is approved as transferer by the template admin.
3. Token is wrapped. 
4. Owner of wrapped token approves marketplace and uses the wrapped token as they would with a normal ERC-1155. 
5. Once the vesting period is up, the current owner of the token can unwrap it. 

4. 
See [Unwrapping](https://hackmd.io/Ucg-sUV_THOWwG3kscpO9g#Unwrapping).
Unwrapping can be sped up for all tokens under the vesting template with `updateTemplate(uint24 templateId, uint96 duration, bool destructionOnly, bool transferOpen)`. A token can also be unwrapped before the vesting period is up by a `templateOperator` using `unwrap(uint256 wrappedTokenId, address owner, uint256 amount)`.

5.
Operators can also unwrap as above. 
The clawback state is only allowed while the token is vesting. Once vested, an operator cannot clawback the token anymore and can only unwrap.